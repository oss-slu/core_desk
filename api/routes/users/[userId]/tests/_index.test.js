import { describe, expect, it } from "vitest";
import request from "supertest";
import { app } from "#index";
import { gt } from "#gt";
import { prisma } from "#prisma";
import { AccountType, LogType } from "@prisma/client";

describe("/[userId]", () => {
  describe("Fetch user profile", () => {
    it("should return a 404 if the user does not exist", async () => {
      const res = await request(app)
        .get(`/api/users/12345`)
        .set(...(await gt()))
        .send();

      expect(res.status).toBe(404);
      expect(res.body).toMatchSnapshot();
    });

    it("should return basic information if the user is not a global admin", async () => {
      const testUser = await prisma.user.create({
        data: {
          firstName: "Test",
          lastName: "User",
          email: "test@example.com",
        },
      });

      const res = await request(app)
        .get(`/api/users/${testUser.id}`)
        .set(...(await gt()))
        .send();

      expect(res.status).toBe(200);

      expect(res.body.user.id).toBe(testUser.id);
      expect(res.body.user).toMatchSnapshot({
        id: expect.any(String),
      });
    });

    it("should not return logs if user is not a global admin", async () => {
      const testUser = await prisma.user.create({
        data: {
          firstName: "Test",
          lastName: "User",
          email: "test@example.com",
          logs: {
            create: [
              {
                type: LogType.USER_LOGIN,
                createdAt: new Date(),
                updatedAt: new Date(),
              },
            ],
          },
        },
      });

      const res = await request(app)
        .get(`/api/users/${testUser.id}`)
        .set(...(await gt()))
        .send();

      expect(res.status).toBe(200);

      expect(res.body.user.id).toBe(testUser.id);
      expect(res.body.user.logs).toBeUndefined();
    });

    it("should return a full profile if the user is a global admin", async () => {
      const testUser = await prisma.user.create({
        data: {
          firstName: "Test",
          lastName: "User",
          email: "targetUser@example.com",
          shops: {
            create: {
              accountType: AccountType.CUSTOMER,
              shopId: (await prisma.shop.findFirst()).id,
            },
          },
        },
      });

      const res = await request(app)
        .get(`/api/users/${testUser.id}`)
        .set(
          ...(await gt({
            ga: true,
          }))
        )
        .send();

      expect(res.status).toBe(200);

      expect(res.body.user.id).toBe(testUser.id);
      expect(res.body.user).toMatchSnapshot({
        id: expect.any(String),
        createdAt: expect.any(String),
        updatedAt: expect.any(String),
        shops: expect.any(Array),
      });

      expect(res.body.user.shopCount).toBe(1);
      expect(res.body.user.jobCount).toBe(0);
      expect(res.body.user.isMe).toBe(false);

      expect(res.body.user.shops[0].shop.name).toBe("TestShop");
      expect(res.body.user.name).toBe("Test User");

      expect(res.body.user._count).toBeUndefined();

      expect(res.body.user.logs).toBeUndefined();
    });

    it("should return logs if includeLogs is true", async () => {
      const testUser = await prisma.user.create({
        data: {
          firstName: "Test",
          lastName: "User",
          email: "targetUser@example.com",
          logs: {
            create: [
              {
                type: LogType.USER_LOGIN,
              },
            ],
          },
        },
      });

      await prisma.user.update({
        where: {
          email: "test@email.com",
        },
        data: {
          admin: true,
        },
      });

      const res = await request(app)
        .get(`/api/users/${testUser.id}?includeLogs=true`)
        .set(...(await gt()))
        .send();

      expect(res.status).toBe(200);

      expect(res.body.user.id).toBe(testUser.id);
      expect(res.body.user).toMatchSnapshot({
        id: expect.any(String),
        createdAt: expect.any(String),
        updatedAt: expect.any(String),
        logs: expect.any(Array),
      });

      expect(res.body.user.logs.length).toBe(1);
      expect(res.body.user.logs[0].type).toBe(LogType.USER_LOGIN);
    });
  });
});
