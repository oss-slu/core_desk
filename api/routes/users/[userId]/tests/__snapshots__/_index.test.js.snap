// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/[userId] > Fetch user profile > should return a 404 if the user does not exist 1`] = `
{
  "message": "User not found",
}
`;

exports[`/[userId] > Fetch user profile > should return a full profile if the user is a global admin 1`] = `
{
  "admin": false,
  "createdAt": Any<String>,
  "email": "targetUser@example.com",
  "firstName": "Test",
  "id": Any<String>,
  "isMe": false,
  "jobCount": 0,
  "jobs": [],
  "lastName": "User",
  "name": "Test User",
  "shopCount": 1,
  "shops": Any<Array>,
  "suspended": false,
  "suspensionReason": null,
  "updatedAt": Any<String>,
}
`;

exports[`/[userId] > Fetch user profile > should return basic information if the user is not a global admin 1`] = `
{
  "firstName": "Test",
  "id": Any<String>,
  "lastName": "User",
  "name": "Test User",
}
`;

exports[`/[userId] > Fetch user profile > should return logs if includeLogs is true 1`] = `
{
  "admin": false,
  "createdAt": Any<String>,
  "email": "targetUser@example.com",
  "firstName": "Test",
  "id": Any<String>,
  "isMe": false,
  "jobCount": 0,
  "jobs": [],
  "lastName": "User",
  "logs": Any<Array>,
  "name": "Test User",
  "shopCount": 0,
  "shops": [],
  "suspended": false,
  "suspensionReason": null,
  "updatedAt": Any<String>,
}
`;
